### ISSUE TITLE
MiniTensor as optional dependency

### ISSUE THREAD

**Issue by mperego** [SNLComputation/Albany#120]
Would it be hard to have minitensor as an optional dependency for Albany? So that, e.g., we do not have to install it when we build Albany/FELIX? This is nor urgent nor essential but it would be a nice feature to have.

**Comment by bartgol** [comment]
That would be fine if MiniTensor was used only in optional packages. I see it in ATO and LCM, but I also see it used in adapt, which makes it hard to make it optional...

**Comment by mperego** [comment]
Yeah, that's the issue. Does MiniTensor  need to be in adapt? Can that part of the code in Albany adapt be moved in another class, or #ifdef guarded? (Adapt is probably the part of Albany core that I understand the least so I'm hesitant to find these answers myself :-) )

**Comment by lxmota** [comment]
@mperego LCM cannot be compiled without MiniTensor. As for Adapt, I took a quick look and it seems that the code that uses MiniTensor there is very old. I'm not even sure that it is used anywhere. As for ATO, my recollection is that it also requires MiniTensor.

So it seems that provided that we can move the code in Adapt or remove it altogether, you could avoid having a dependency on MiniTensor for FELIX.

**Comment by mperego** [comment]
Thanks @lxmota.  If you think that the code in Albany adapt that uses MiniTensor might be obsolete then it probably makes sense anyway to have a look at it and clean it up. As I said it is not urgent.

**Comment by bgranzow** [comment]
I'm pretty sure the adapt code only uses MiniTensor for reference configuration updates (RCU). It looks like there is at least one [test](http://my.cdash.org/testSummary.php?project=688&name=SCOREC_RCU_SquareStabilized_Tpetra_SERIAL&date=2017-04-28) that still runs with RCU enabled.

It was my understanding that the RCU code was intended to be a prototype for further research codes with adaptivity. It would make sense to check with @gahansen and @ibaned to see if removing this code is the right thing to do. 

It may be possible to make RCU optional via CMake in the adapt directory. Some care would need to be taken in this approach, as at least one RCU variable is utilized outside of the adapt code:
https://github.com/gahansen/Albany/blob/master/src/Albany_Application.cpp#L197
https://github.com/gahansen/Albany/blob/master/src/problems/Albany_ProblemFactory.hpp#L63

**Comment by ibaned** [comment]
Thanks @bgranzow . Finally I understand whats going on here.

I think the parts of RCU that use MiniTensor are likely quite specific to LCM anyways (they're managing dispacements and deformation gradients), so its not like they're used even in adaptive FELIX runs. Maybe we'll just `#ifdef` that code out and throw an error message if its called without MiniTensor enabled.

**Comment by mperego** [comment]
@bgranzow  @ibaned Is any of you willing to address this issue, maybe cleaning up the parts of the code that are not used any more since we already are in a cleaning mood ;-).

**Comment by bgranzow** [comment]
Hi @mperego, I think some higher-level decisions need to be made before we can commit to this course of action. Namely, what we intend to do to resolve this issue:
https://github.com/gahansen/Albany/issues/199

**Comment by lxmota** [comment]
@mperego With the forking of LCM and removal of other packages, MiniTensor is no longer needed. I can be safely removed, I think.
